nat/sort.go

```diff
@@ -26,3 +29,55 @@ func Sort(ports []Port, predicate func(i, j Port) bool) {
 	s := &portSorter{ports, predicate}
 	sort.Sort(s)
 }
+
+type portMapEntry struct {
+	port    Port
+	binding PortBinding
+}
+
+type portMapSorter []portMapEntry
+
+func (s portMapSorter) Len() int      { return len(s) }
+func (s portMapSorter) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s portMapSorter) Less(i, j int) bool {
+	pi, pj := toInt(string(s[i].port)), toInt(string(s[j].port))
+	hpi, hpj := toInt(s[i].binding.HostPort), toInt(s[j].binding.HostPort)
+	return (hpi > hpj) || (pi > pj)
+}
+
+// SortPortMap sorts the list of ports and their respected mapping. The ports
+// will explicit HostPort will be placed first.
+func SortPortMap(ports []Port, bindings PortMap) {
+	s := portMapSorter{}
+	for _, p := range ports {
```

Comment:
So the idea is to sort the list of ports and port bindings so that port which has explicit port bindings comes first. This put sport-binding combinations into a list, sort them and then rewrite the original list / bindings.

I hope that's clearer :D